File: javascript/heapSort.js
let array = [];
let n = Math.floor(Math.random() * 1000000);

for (let i = 0; i < n; i++) {
	array[i] = Math.floor(Math.random() * 1000000000);
}
console.log("Unsorted Array: " + array);

let heapSize;

function heapSort(A) {
	buildMaxHeap(A);
	for (let i = A.length - 1; i >= 1; i--) {
		let temp = A[0];
		A[0] = A[i];
		A[i] = temp;
		heapSize = heapSize - 1;
		maxHeapify(A, 0);
	}
}

function buildMaxHeap(A) {
	heapSize = A.length;
	for (let i = Math.floor(A.length / 2) - 1; i >= 0; i--) {
		maxHeapify(A, i);
	}
}

function maxHeapify(A, i) {
	let l = LEFT(i);
	let r = RIGHT(i);
	let max = i;

	if (l < heapSize && A[l] > A[max]) {
		max = l;
	}
	if (r < heapSize && A[r] > A[max]) {
		max = r;
	}
	if (max != i) {
		let temp = A[i];
		A[i] = A[max];
		A[max] = temp;
		maxHeapify(A, max);
	}
}

function LEFT(i) {
	return 2 * i + 1;
}

function RIGHT(i) {
	return 2 * i + 2;
}

let startTime = performance.now();
heapSort(array);
let endTime = performance.now();

console.log("Sorted Array: " + array);
console.log("Execution Time: " + (endTime - startTime) + " ms!");

File: python/heapSort.py
import time
import random
import math

def left(i):
	return 2 * i + 1

def right(i):
	return 2 * i + 2

def maxHeapify (A, i, heapSize):
	l = left(i)
	r = right(i)
	maxIndex = i

	if l < heapSize and A[l] > A[maxIndex]:
		maxIndex = l
	
	if r < heapSize and A[r] > A[maxIndex]: 
		maxIndex = r

	if maxIndex != i:
		A[i], A[maxIndex] = A[maxIndex], A[i]
		maxHeapify(A, maxIndex, heapSize)

def buildMaxHeap (A):
	heapSize = len(A)
	for i in range(heapSize//2 - 1, -1, -1):
		maxHeapify(A, i, heapSize)

def heapSort (A):
	heapSize = len(A)
	buildMaxHeap(A)
	for i in range(len(A) - 1, 0, -1):
		A[i], A[0] = A[0], A[i]
		heapSize -= 1
		maxHeapify(A, 0, heapSize)


A = [random.randint(1, 10) for _ in range(10)]

print("Unsorted array:", A)

startTime = time.time()

heapSort(A)

endTime = time.time()
executionTime = (endTime - startTime)

print("Sorted array:", A)
print(f"Execution time: {executionTime:.6f} seconds")